### PR TITLE
[ROCm] Enable unique_op for ROCm

### PR DESCRIPTION
**Pull request by jayfurmanek**
This PR enables the unique_op_gpu (int32 and int64) for ROCM on AMD GPUs.

@cheshire @chsigg @deven-amd

### CODE REVIEW

File: tensorflow/core/kernels/unique_op_gpu.cu.h [hunk 1]
```diff
@@ -292,9 +292,10 @@ class UniqueOpGPU : public AsyncOpKernel {
     using namespace unique_op_gpu;
 
     // Create a fancy input iterator to indicate segment boundaries.
+    gpuprim::CountingInputIterator<TIndex> counting_iter(0);
     gpuprim::TransformInputIterator<TIndex, SegmentIndicatorFunctor<T, TIndex>,
                                     gpuprim::CountingInputIterator<TIndex>>
-        segment_indicator_iter(0, {sorted_input_ptr});
+        segment_indicator_iter(counting_iter, {sorted_input_ptr});
```
Comment (cheshire) [inline tensorflow/core/kernels/unique_op_gpu.cu.h:298]: Maybe initialize inline for readability?

File: tensorflow/core/kernels/unique_op_gpu.cu.h [hunk 2]
```diff
@@ -331,7 +332,11 @@ class UniqueOpGPU : public AsyncOpKernel {
       const GPUDevice& device = context->eigen_gpu_device();
       int64 uniq_size = (*last_idx_host.data()) + 1;
 
+#if GOOGLE_CUDA
       se::cuda::ScopedActivateExecutorContext scoped_activation{
+#else
+      se::gpu::ScopedActivateExecutorContext scoped_activation{
```
Comment (cheshire) [inline tensorflow/core/kernels/unique_op_gpu.cu.h:338]: Can we use se::gpu:: for both?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/unique_op_gpu.cu.h (hunk 1 above)
Comment (jayfurmanek) [reply tensorflow/core/kernels/unique_op_gpu.cu.h:298]: It looks like most/all of the other locations in the codebase where these iterators are used are structured like this where the counting iterator is created/initialized up front and then passed in to a transform iterator. Maybe it's best to match?

File: tensorflow/core/kernels/unique_op_gpu.cu.h (hunk 2 above)
Comment (jayfurmanek) [reply tensorflow/core/kernels/unique_op_gpu.cu.h:338]: Yea, this was probably overly careful. I'll test this as se::gpu on a NV machine to make sure.

**Review by cheshire — APPROVED** [review]

**Comment by cheshire** [comment]
Should some of the tests be enabled for ROCm as well?

**Comment by jayfurmanek** [comment]
Thanks @cheshire !  
It looks like the unique_op_test (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/unique_op_test.cc) is cpu only and is therefore already enabled across the board. Was there any other test you had in mind?

**Comment by cheshire** [comment]
If there's no GPU test you don't have to write one, but ideally, there really should be.

**Comment by jayfurmanek** [comment]
I agree! I'll take a look at the existing test, maybe there is an easy way to have it also run on GPU(s).

**Pull request merged**